in_toto/runlib.py

```diff
@@ -571,31 +574,67 @@ def in_toto_run(name, material_list, product_list, link_cmd_args,
-    LOG.info("Signing link metadata using passed key...")
-    signature = link_metadata.sign(signing_key)
+    signer = None
+    if signing_key:
+      LOG.info("Signing link metadata using passed key...")
+      signer = SSlibSigner(signing_key)
 
-  elif gpg_keyid:
-    LOG.info("Signing link metadata using passed GPG keyid...")
-    signature = link_metadata.sign_gpg(gpg_keyid, gpg_home=gpg_home)
+    elif gpg_keyid:
+      LOG.info("Signing link metadata using passed GPG keyid...")
+      signer = GPGSigner(keyid=gpg_keyid, homedir=gpg_home)
 
-  elif gpg_use_default:
-    LOG.info("Signing link metadata using default GPG key ...")
-    signature = link_metadata.sign_gpg(gpg_keyid=None, gpg_home=gpg_home)
+    elif gpg_use_default:
+      LOG.info("Signing link metadata using default GPG key ...")
+      signer = GPGSigner(keyid=None, homedir=gpg_home)
 
-  # We need the signature's keyid to write the link to keyid infix'ed filename
-  if signature:
-    signing_keyid = signature["keyid"]
-    filename = FILENAME_FORMAT.format(step_name=name, keyid=signing_keyid)
+    if signer:
+      signature = link_metadata.sign(signer)
+
+      signing_keyid = signature.keyid
+      filename = FILENAME_FORMAT.format(step_name=name, keyid=signing_keyid)
+
+      if metadata_directory is not None:
+        filename = os.path.join(metadata_directory, filename)
+
+      LOG.info("Storing link metadata to '{}'...".format(filename))
+      metadata_dict = link_metadata.to_dict()
+
+      indent = None if compact_json else 1
+      separators = (',', ':') if compact_json else (',', ': ')
+
+      with open(filename, 'w', encoding='utf8') as fp:
+        json.dump(metadata_dict, fp, indent=indent, separators=separators)
+
+  else:
+    link_metadata = Metablock(signed=link, compact_json=compact_json)
+
+    signature = None
+    if signing_key:
+      LOG.info("Signing link metadata using passed key...")
+      signature = link_metadata.sign(signing_key)
+
+    elif gpg_keyid:
+      LOG.info("Signing link metadata using passed GPG keyid...")
+      signature = link_metadata.sign_gpg(gpg_keyid, gpg_home=gpg_home)
+
+    elif gpg_use_default:
+      LOG.info("Signing link metadata using default GPG key ...")
+      signature = link_metadata.sign_gpg(gpg_keyid=None, gpg_home=gpg_home)
+
+    # We need the signature's keyid to write the link to keyid infix'ed
+    # filename
+    if signature:
+      signing_keyid = signature["keyid"]
+      filename = FILENAME_FORMAT.format(step_name=name, keyid=signing_keyid)
 
-    if metadata_directory is not None:
-      filename = os.path.join(metadata_directory, filename)
+      if metadata_directory is not None:
+        filename = os.path.join(metadata_directory, filename)
 
-    LOG.info("Storing link metadata to '{}'...".format(filename))
-    link_metadata.dump(filename)
+      LOG.info("Storing link metadata to '{}'...".format(filename))
+      link_metadata.dump(filename)
 
   return link_metadata
 
```

Comment:
I think this should be straightforward to implement for `in_toto_record_*` and `in_toto_mock` as well!

Reply:
I suggest to hold off porting this feature to `in_toto_record` and `in_toto_mock` until the metadata architecture has crystallized (see https://github.com/in-toto/in-toto/pull/503#pullrequestreview-1058361691)